File: 145_Longest Palindromic Substring.cpp
/*
145. Longest Palindromic Substring
Find the longest palindromic substring by expanding around centers (O(n^2)).
*/
#include <bits/stdc++.h>
using namespace std;

string longestPalindrome(string s) {
    int n = s.size();
    if (n < 2) return s;
    int start = 0, maxLen = 1;
    for (int i = 0; i < n; ++i) {
        // odd
        int l = i, r = i;
        while (l >= 0 && r < n && s[l] == s[r]) {
            if (r - l + 1 > maxLen) { start = l; maxLen = r - l + 1; }
            --l; ++r;
        }
        // even
        l = i; r = i + 1;
        while (l >= 0 && r < n && s[l] == s[r]) {
            if (r - l + 1 > maxLen) { start = l; maxLen = r - l + 1; }
            --l; ++r;
        }
    }
    return s.substr(start, maxLen);
}

int main() {
    cout << longestPalindrome("babad") << "\n"; // "bab" or "aba"
    cout << longestPalindrome("cbbd") << "\n"; // "bb"
    return 0;
}

// Time Complexity: O(n^2)
// Space Complexity: O(1)

File: 145_Longest Palindromic Substring.py
"""
145. Longest Palindromic Substring
Expand around each center (odd and even) to find longest palindrome.
"""
from typing import Tuple

def longest_palindrome(s: str) -> str:
    n = len(s)
    if n < 2: return s
    start, maxlen = 0, 1
    for i in range(n):
        # odd
        l, r = i, i
        while l >= 0 and r < n and s[l] == s[r]:
            if r - l + 1 > maxlen:
                start, maxlen = l, r - l + 1
            l -= 1; r += 1
        # even
        l, r = i, i+1
        while l >= 0 and r < n and s[l] == s[r]:
            if r - l + 1 > maxlen:
                start, maxlen = l, r - l + 1
            l -= 1; r += 1
    return s[start:start+maxlen]

if __name__ == '__main__':
    print(longest_palindrome('babad'))
    print(longest_palindrome('cbbd'))

# Time Complexity: O(n^2)
# Space Complexity: O(1)
